introduce module HTTP and generic request function

royneary opened the pull request:
While trying out ``fmlib_browser`` I realized that there is no easy way to send HTTP requests with JSON bodies.

``http_json`` takes a string as its ``body`` argument. The ``Value`` module allows constructing JSON values, but has no ``to_string`` function. So to send JSON bodies right now I have to construct JSON strings manually.

I see two alternatives for solving this:

1. Add a ``to_string`` function to module ``Value``

2. Change ``http_json`` so it accepts a ``Value.t option`` as its ``body`` argument

I implemented 2. because it gives us the more convenient and type-safe API. Also, I would argue that sending *and* receiving JSON is a very common use case (much more common that sending text and receiving JSON).

It is a breaking change though.

hbr commented:
Hello Christian,

thanks for the pull request. I'm just analyzing it and have not yet come to a conclusion. Some thoughts:

1. I agree that the main method to interchange structured information in http-requests is to use json encoding. Therefore I have made to methods `http_text` which receives the http resonse as an unstructured text and `http_json` which expects the response in json encoding. I have not included the possibility to send the body of the request json encoded as well.

2. Your proposal does not work as expected for `http_text` because the user adds the unstructured text `"hello"` as body and it gets a request sent as `"\"hello\""` (the json encoding of a string is the quoted string).

3. In order to send a http request in javascript we need an `XmlHttpRequest` object. The `send` method of the object accepts a `Blob`, `TypedArray`, ... or a `string` (according to https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/send). In the `js` library of `fmlib` I have currently only implemented the sending with a string. Future extension of `js` can include to possibility to send blobs, etc as alternatives to send strings. Sending objects of type `Base.Value.t` makes sending of an unstructured string impossible (see 2.).

4. In `js` the module `Base.Value` has the function `stringify` with the signature `t -> t option`. It returns a `Value.t option` because the object might contain objects which are not serializable (e.g. functions). However an object of type `Value.t` in `fmlib_browser` is guaranteed to be serializable. Therefore it would be absolutely type safe to add a function `to_json` which returns a string (and not a string option) given an object of type `Value.t` in `fmlib_browser`.

5. The encoding of the request body and the encoding of the response text seem to be independent. I can imagine situations where you expect unstructured text as a response but the body must be structured text in json encoding. It is possible to expect json encoded response but the body is unstructured (probably less likely than the previous case).

What are your thoughts? I have not yet come to a conclusion. Any comments are welcome.

Regards
Helmut

royneary commented:
Hi Helmut,

thank you for your kind reply. You're completely right, I messed up the string representation in ``http_text``. I also understand that the API should be generic and extensible.

I looked at [elm/http](https://package.elm-lang.org/packages/elm/http/latest/) for some inspiration. There they have two types ``Body`` and ``Expect`` to model different kinds of HTTP payloads.
I tried to implement the same thing for ``fmlib_browser`` by introducing a generic function ``http`` and the two modules ``Http_body`` and ``Http_expect``. Please tell me if you don't like the naming or how the modules are organized, I'm pretty new to OCaml and its module system.
This approach should allow us to extend the API later (e.g. file uploads) without breaking changes if I'm not mistaken.

I changed the return type of ``Base.Value.stringify`` to ``string option`` because I couldn't figure out how to use it otherwise.

I kept ``http_text`` and ``http_json`` for now and implemented them in terms of the generic function.

There is no documentation yet, I can write it later. What do you think?

royneary commented:
(I messed up something and the PR got closed automatically)

I think I implemented what you described.

In ``Http.request`` we could add ``Content-Type`` and ``Accept`` headers automatically based what kind of body and expect values are passed in (similar to ``elm/http``). But maybe that's for another PR?

hbr commented:
Hi Christian,

I have analyzed your PR, merged it into the branch http.

Your PR is now much simpler and more straightforward than the previous one and was therefore easy to analyze.

I have made one modification. I have eliminated the module `Task.Http.Error` and put the error type directly into `Task.Http`. Thanks for your work.
